### DISCUSSION TITLE
Date Based filteration fails always

### DISCUSSION THREAD

**Discussion by rootandleaves** [prisma/prisma#4193] [Q&A] [answered]
I am trying to filter coupons based on expire date and availability but following condition always fails

```js
await Context.DataBase.discount.findMany({ skip, take, 'where': { 'AND': [{{ 'expireAt': { 'lt': expireAt ? new Date(expireAt).toISOString() : new Date().toISOString() } }, { 'availableAt': { 'gt': availableAt ? new Date(availableAt).toISOString() : new Date().toISOString() } }] } })
```

and in my database coupons are stored with expireAt and availableAt in ISO String format. what i am doing wrong here ?.

**Comment by ryands17** [comment]
@rootandleaves What is the datatype of the date fields? Also could you share your schema and prisma versions so that I can reproduce.

**Comment by rootandleaves** [comment]
Following is my schema.

```
model TruckpeDiscount {
  id                     String                   @id @default(cuid())
  createdAt              DateTime                 @default(now())
  updatedAt              DateTime                 @default(now())
  availableAt            DateTime                 @default(now())
  expireAt               DateTime                 @default(now())
  code                   String                   @unique
  canUse                 account_type
  discount               Int
  description            String
  AccountTruckpeDiscount AccountTruckpeDiscount[] @relation("TruckpeDiscountTo_AccountTruckpeDiscount")
}
````

**Reply by ryands17** [reply, marked as answer]
The query works fine for me with random data inserted as per your schema. Also `toISOString()` is not required as Prisma directly accepts Date objects.


```prisma
model TruckpeDiscount {
  id          String   @id @default(cuid())
  createdAt   DateTime @default(now())
  updatedAt   DateTime @default(now())
  availableAt DateTime @default(now())
  expireAt    DateTime @default(now())
  code        String   @unique
  discount    Int
  description String
}
```

I used the following query:

```ts
  await prisma.truckpeDiscount.findMany({
    where: {
      AND: [
        {
          expireAt: {
            lt: expireAt ? new Date(expireAt) : new Date(),
          },
        },
        {
          availableAt: {
            gt: availableAt ? new Date(availableAt) : new Date(),
          },
        },
      ],
    },
  })
```

**Reply by rootandleaves** [reply]
ok...One more thing is there any way to insert custom date when Prisma returns it...for example...when Prisma queries FindMany(...) and returns array [] of data...is it Possible to insert my own custom key pair in this array to avoid loop on FindMany result ?

**Reply by rootandleaves** [reply]
Can you query this for me..this query is not working either.

`
await Context.DataBase.discount.findMany({ 
    'where': { 'AND': [
            {
              'availableAt': {
                'lte': new Date(Moment().add( 3, 'days').toISOString()),
                'gte': availableAt ? new Date(availableAt) : new Date()
              }
            }
          ] } })
`

**Reply by ryands17** [reply]
> One more thing is there any way to insert custom date when Prisma returns it...for example...when Prisma queries FindMany(...) and returns array [] of data...is it Possible to insert my own custom key pair in this array to avoid loop on FindMany result ?

You would only be able to insert custom keys by looping on the result as `findMany` always returns an array.

As for the query, I tried it like this:

```ts
const availableAt = new Date()

  await prisma.truckpeDiscount.findMany({
    where: {
      AND: [
        {
          availableAt: {
            lte: Moment().add(3, 'days').toDate(),
            gte: availableAt ? new Date(availableAt) : new Date(),
          },
        },
      ],
    },
  })
```

And this works too. What is the error that you are getting on running the query? Could you share that so I can get a better idea.

**Reply by rootandleaves** [reply]
Your Reply: You would only be able to insert custom keys by looping on the result as findMany always returns an array.

Yes i can do it but if 50000 users used Given Discount Coupon. than i have to loop and check for any given particular user name "X" in it...every time query is hit..Which is totally waste according to me..thats why i think you guys should allow to add Object on query return..instead of looping on huge data set..Just suggestion..

**Reply by rootandleaves** [reply]
it worked..now sorry for making mistake from my end..

**Reply by ryands17** [reply]
> Yes i can do it but if 50000 users used Given Discount Coupon. than i have to loop and check for any given particular user name "X" in it...every time query is hit.

I think this should be handled on the database end, by querying it with Prisma. `findMany` also can search via a name so you can pass that in your query to make things faster.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
